File: 29Items-Projects/7Algorithms/7-String-algorithms/python/string_algorithms/algorithms/aho_corasick.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field


@dataclass(frozen=True)
class MatchResult:
    algorithm: str
    pattern: str
    start_index: int


@dataclass
class _Node:
    children: dict[str, "_Node"] = field(default_factory=dict)
    outputs: list[int] = field(default_factory=list)
    failure: "_Node | None" = None


class AhoCorasick:
    def __init__(self, patterns: list[str]) -> None:
        if not patterns or any(pattern == "" for pattern in patterns):
            raise ValueError("patterns must contain at least one non-empty value")
        self._patterns = list(patterns)
        self._root = _Node()
        self._build_trie()
        self._build_failure_links()

    def search(self, text: str) -> list[MatchResult]:
        matches: list[MatchResult] = []
        current = self._root

        for index, char in enumerate(text):
            while current is not self._root and char not in current.children:
                current = current.failure or self._root

            current = current.children.get(char, self._root)

            for pattern_index in current.outputs:
                pattern = self._patterns[pattern_index]
                matches.append(
                    MatchResult("Aho-Corasick", pattern, index - len(pattern) + 1)
                )

        return matches

    def _build_trie(self) -> None:
        for pattern_index, pattern in enumerate(self._patterns):
            current = self._root
            for char in pattern:
                current = current.children.setdefault(char, _Node())
            current.outputs.append(pattern_index)

    def _build_failure_links(self) -> None:
        self._root.failure = self._root
        queue: deque[_Node] = deque()

        for child in self._root.children.values():
            child.failure = self._root
            queue.append(child)

        while queue:
            current = queue.popleft()
            for char, child in current.children.items():
                fallback = current.failure or self._root
                while fallback is not self._root and char not in fallback.children:
                    fallback = fallback.failure or self._root

                child.failure = fallback.children.get(char, self._root)
                child.outputs.extend(child.failure.outputs)
                queue.append(child)

File: 29Items-Projects/7Algorithms/7-String-algorithms/src/main/java/com/stringalgorithms/algorithms/AhoCorasick.java
package com.stringalgorithms.algorithms;

import com.stringalgorithms.model.MatchResult;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Queue;

public final class AhoCorasick {
    private final Node root = new Node();
    private final List<String> patterns;

    public AhoCorasick(List<String> patterns) {
        Objects.requireNonNull(patterns, "patterns must not be null");
        if (patterns.isEmpty() || patterns.stream().anyMatch(String::isEmpty)) {
            throw new IllegalArgumentException("patterns must contain at least one non-empty value");
        }
        this.patterns = List.copyOf(patterns);
        buildTrie();
        buildFailureLinks();
    }

    public List<MatchResult> search(String text) {
        Objects.requireNonNull(text, "text must not be null");
        List<MatchResult> matches = new ArrayList<>();
        Node current = root;

        for (int index = 0; index < text.length(); index++) {
            char ch = text.charAt(index);
            while (current != root && !current.children.containsKey(ch)) {
                current = current.failure;
            }
            current = current.children.getOrDefault(ch, root);

            for (int patternIndex : current.outputs) {
                String pattern = patterns.get(patternIndex);
                matches.add(new MatchResult("Aho-Corasick", pattern, index - pattern.length() + 1));
            }
        }

        return matches;
    }

    private void buildTrie() {
        for (int patternIndex = 0; patternIndex < patterns.size(); patternIndex++) {
            Node current = root;
            for (char ch : patterns.get(patternIndex).toCharArray()) {
                current = current.children.computeIfAbsent(ch, ignored -> new Node());
            }
            current.outputs.add(patternIndex);
        }
    }

    private void buildFailureLinks() {
        Queue<Node> queue = new ArrayDeque<>();
        root.failure = root;

        for (Node child : root.children.values()) {
            child.failure = root;
            queue.add(child);
        }

        while (!queue.isEmpty()) {
            Node current = queue.remove();
            for (Map.Entry<Character, Node> entry : current.children.entrySet()) {
                char ch = entry.getKey();
                Node child = entry.getValue();
                Node fallback = current.failure;

                while (fallback != root && !fallback.children.containsKey(ch)) {
                    fallback = fallback.failure;
                }

                child.failure = fallback.children.getOrDefault(ch, root);
                child.outputs.addAll(child.failure.outputs);
                queue.add(child);
            }
        }
    }

    private static final class Node {
        private final Map<Character, Node> children = new HashMap<>();
        private final List<Integer> outputs = new ArrayList<>();
        private Node failure;
    }
}
